### PR TITLE
Restore where-filter null handling to match earlier versions (<3.2)

### PR DESCRIPTION
**Pull request by chuckhoupt**
Summary of full bug report #6038:

> Before Jekyll 3.2, the `where` filter had the useful ability to select objects based on a missing or empty keys (`where: "key", nil`). The new array-support feature (pull #4555) changed this behavior causing problems for sites upgrading from 3.1.

This pull request tests and fixes the bug introduced in pull #4555. The core problem was that `Array(nil)` returns `[]`, rather then `[nil]`, so the fix uses an immediate array constructor to preserve nils.

### CODE REVIEW

**Comment by pathawks** [comment]
> fixes the bug 

Just so we're clear, was this previous behavior documented, or is this just a case where the undefined behavior changed?

Will this PR break somebody's site that depends on the _new_ behavior?

**Comment by chuckhoupt** [comment]
The previous behavior was undocumented/untested, so it is a change to undefined behavior. It will break sites that depend on the new behavior (often by subtly changing output). So I see that my argument for restoring the old behavior is a weak appeal to "tradition", and maybe it is too late to restore it after 10 months.

Possibly a more conservative alternative would be to simply make the undefined areas illegal, so an error or warning is thrown? That would give a clear signal that the old behavior has changed and new sites shouldn't rely on it.

I only noticed the change by running a diff of a site while upgrading from 3.1 to 3.4 -- it could easily have been missed. A warning would have helped.

File: test/test_filters.rb [hunk 1]
```diff
@@ -655,6 +655,17 @@ def to_liquid
...
       end
 
+      should "filter properties with nil values and undefined properties appropriately" do
+        objects = [
+          {},
+          { "a" => nil },
+          { "a" => "" },
+        ]
+        assert_equal 3, @filter.where(objects, "a", nil).length
+        assert_equal 3, @filter.where(objects, "a", "").length
```
Comment (parkr) [inline test/test_filters.rb]: ??

File: lib/jekyll/filters.rb [hunk 2]
```diff
@@ -221,7 +221,9 @@ def where(input, property, value)
       return input unless input.respond_to?(:select)
       input = input.values if input.is_a?(Hash)
       input.select do |object|
-        Array(item_property(object, property)).map(&:to_s).include?(value.to_s)
+        item_value = item_property(object, property)
+        item_value = [item_value] unless item_value.respond_to?("map")
```
Comment (parkr) [inline lib/jekyll/filters.rb]: This should be `:map` instead of `"map"`.

File: lib/jekyll/filters.rb [hunk 3]
```diff
@@ -221,7 +221,9 @@ def where(input, property, value)
...
+        item_value.map(&:to_s).include?(value.to_s)
```
Comment (parkr) [inline lib/jekyll/filters.rb]: Could we special-case `value`? Then we could do

```ruby
if value.nil?
  item_value.any?(&:nil?)
else
  item_value.map(&:to_s).include?(value.to_s)
end
```

File: lib/jekyll/filters.rb (hunk 2 above)
Comment (chuckhoupt) [inline lib/jekyll/filters.rb]: Fixed.

File: test/test_filters.rb (hunk 1 above)
Comment (chuckhoupt) [inline test/test_filters.rb]: I improved the test description.

File: lib/jekyll/filters.rb (hunk 3 above)
Comment (chuckhoupt) [inline lib/jekyll/filters.rb]: The problem with special-casing nil is that it would break the existing behavior. `where` has always treated nil and empty-string (as well as nonexistent properties) as identical because they all stringify to the empty-string. The array-support added by #4555 continued this tradition.

I've added some more tests to make the current handling of nil/empty-string in list properties explicit. A special case for nil will break these tests.

**Comment by chuckhoupt** [comment]
I realize that `where`'s identical handling of nil and empty-string are a bit odd, but I think changing it may break existing sites.

**Comment by pathawks** [comment]
How do we want to approach this in `4.0`?
